blockstore/api.go

```diff
@@ -31,31 +40,44 @@ func (a *apiBlockstore) DeleteBlock(context.Context, cid.Cid) error {
 }
 
 func (a *apiBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	if a.cache.Contains(c) {
+		return true, nil
+	}
 	return a.api.ChainHasObj(ctx, c)
 }
 
 func (a *apiBlockstore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	if bb, ok := a.cache.Get(c); ok {
+		return blocks.NewBlockWithCid(bb, c)
+	}
 	bb, err := a.api.ChainReadObj(ctx, c)
 	if err != nil {
 		return nil, err
 	}
+	a.cache.Add(c, bb)
```

Comment:
Use `ContainsOrAdd` for better performance in a highly concurrent environment. That would reduce unnecessary churn in LRU.